content/blog/2022-05-05-e2e-cv-pt3.md

```diff
@@ -0,0 +1,232 @@
+2. submit your ML training job to it
+3. get the results back
+4. automatically shut down the VM without having to worry about excessive cloud
+   bills
+
+then the answer is yes if you use the [CML library](https://cml.dev/) in
+combination with CI/CD tools like GitHub Actions or GitLab CI/CD.
+
+![Continuous Integration and Deployment for Machine Learning](https://dvc.org/static/300c88b3b1b5f65753629d661cc916e5/2e49e/cicd4ml.png)
+
+We've configured three
+[workflow files](https://github.com/iterative/magnetic-tiles-defect/tree/main/.github/workflows)
+for GitHub Actions, each of which corresponds to a particular stage depending on
+the project's lifecycle we are in:
+
+1. [Workflow for experimentation and hyperparameter tuning](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/1-experiment.yaml)
+
+   ![Workflow for experimentation and hyperparameter tuning](/uploads/images/2022-05-05/workflow_exp.png '=800')
+
+   In this stage, we'll be working on an experimentation git branch:
+   experimenting with data preprocessing, changing model architecture, tuning
+   hyperparameters, etc. Once we think our experiment is ready to be run, we'll
+   push our changes to a remote repository (in this case, GitHub). This push
+   will trigger a CI/CD job in GitHub Actions, which in turn will:
+
+   1. provision an EC2 virtual machine with a GPU in AWS:
+
+      ```yaml
+      ...
+      - name: deploy
+              env:
+                REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
+              run: |
+                cml runner \
+                    --cloud=aws \
+                    --cloud-region=us-east-1 \
+                    --cloud-type=g4dn.xlarge \
+                    --labels=cml-runner
+      ...
+      ```
+
+   2. deploy our experiment branch to a docker container on this machine:
+
+      ```yaml
+      ---
+      train-model:
+        needs: deploy-runner
+        runs-on: [self-hosted, cml-runner]
+        container:
+          image: iterativeai/cml:0-dvc2-base1
+          options: --gpus all
+        environment: cloud
+        permissions:
+          contents: read
+          id-token: write
+        steps:
+          - uses: actions/checkout@v2
+          - uses: actions/setup-python@v2
+            with:
+              python-version: '3.9'
+      ```
+
+   3. rerun the entire DVC pipeline & push metrics back to GitHub;
+
+      ```yaml
+      ...
+      - name: dvc-repro-cml
+              env:
+                REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
+              run: |
+                # Install dependencies
+                pipenv install --skip-lock
+                pipenv run dvc pull
+                pipenv run dvc exp run
+                pipenv run dvc push
+      ...
+      ```
+
+   4. open a pull request and post a report to it that contains a table with
+      metrics and model outputs (segmentation masks) on a few test images:
+
+      ```yaml
```

Comment:
Maybe use another syntax highlighter? The block looks too gray and comments are not distinguished.

```suggestion
      ```bash
```